cura/Settings/MachineManager.py

```diff
@@ -729,22 +729,41 @@ def renameMachine(self, machine_id: str, new_name: str) -> None:
     @pyqtSlot(str)
     def removeMachine(self, machine_id: str) -> None:
         # If the machine that is being removed is the currently active machine, set another machine as the active machine.
-        activate_new_machine = (self._global_container_stack and self._global_container_stack.getId() == machine_id)
+        is_currently_active_machine = (self._global_container_stack and self._global_container_stack.getId() == machine_id)
+
+        container_registry = self._application.getContainerRegistry()
+        # Get all the extruder stacks that belong to this machine, need to remove them later as well.
+        if is_currently_active_machine:
+            global_stack = self._global_container_stack
+            extruder_stack_list = list(self._global_container_stack.extruders.values())
+        else:
+            machine_stack_metadata = container_registry.findContainerStacksMetadata(id = machine_id, type = "machine")
+            if not machine_stack_metadata:
+                Logger.log("i", "Machine with ID [%s] not found. Cannot remove it.", machine_id)
+                return
+            global_stack = machine_stack_metadata
+            extruder_stack_list = container_registry.findContainerStacks(machine = machine_id, type = "extruder_train")
 
         # activate a new machine before removing a machine because this is safer
-        if activate_new_machine:
-            machine_stacks = CuraContainerRegistry.getInstance().findContainerStacksMetadata(type = "machine")
+        if is_currently_active_machine:
+            machine_stacks = container_registry.findContainerStacksMetadata(type = "machine")
             other_machine_stacks = [s for s in machine_stacks if s["id"] != machine_id]
             if other_machine_stacks:
                 self.setActiveMachine(other_machine_stacks[0]["id"])
 
         metadata = CuraContainerRegistry.getInstance().findContainerStacksMetadata(id = machine_id)[0]
         network_key = metadata["um_network_key"] if "um_network_key" in metadata else None
-        ExtruderManager.getInstance().removeMachineExtruders(machine_id)
-        containers = CuraContainerRegistry.getInstance().findInstanceContainersMetadata(type = "user", machine = machine_id)
+
+        for extruder_stack in extruder_stack_list:
+            container_registry.removeContainer(extruder_stack.userChanges.getId())
+            container_registry.removeContainer(extruder_stack.definitionChanges.getId())
+            container_registry.removeContainer(extruder_stack.getId())
+
+        container_registry.removeContainer(global_stack.definitionChanges.getId())
+        containers = container_registry.findInstanceContainersMetadata(type = "user", machine = machine_id)
```

Comment:
Can't you just remove `global_stack.userChanges.getId()` from the registry? Why this db query?